src/txn/txn.c

```diff
@@ -177,16 +177,21 @@ __wt_txn_get_snapshot(WT_SESSION_IMPL *session)
     /*
      * Include the checkpoint transaction, if one is running: we should ignore any uncommitted
      * changes the checkpoint has written to the metadata. We don't have to keep the checkpoint's
-     * changes pinned so don't including it in the published pinned ID.
+     * changes pinned so don't go including it in the published pinned ID.
+     *
+     * We can assume that if a function calls without intention to publish then it is the special
+     * case of checkpoint calling it twice. In which case do not include the checkpoint id.
      */
     if ((id = txn_global->checkpoint_txn_shared.id) != WT_TXN_NONE) {
-        txn->snapshot[n++] = id;
-        txn_shared->metadata_pinned = id;
+        if (txn->id != txn_global->checkpoint_txn_shared.id)
```

Comment:
Can you please change this to `if (id != txn->id)` -- in other words, don't go to `txn_global` again after we copy the ID?

(fwiw, this may not matter as Keith says the C standard allows the compiler to ignore local variables, but at least when I read this code, the idea is that we're using the local variable to indicate that we want to read the shared info once, and then don't expect the value we've read to change under us).